### PR TITLE
support join reorder.

### PR DESCRIPTION
**Pull request by hanfei19910905**
I implement a simple join reorder algo.
@coocood @shenli @zimulala PTAL

### CODE REVIEW

File: plan/join_reorder.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+package plan
+
+import (
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+)
+
+func tryToGetJoinGroups(j *Join) ([]LogicalPlan, bool) {
```
Comment (coocood) [inline plan/join_reorder.go]: It returns a slice of Plan in a single group, so the name should be 'tryToGetJoinGroup'.

File: plan/join_reorder.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	join.initID()
+	join.SetChildren(lChild, rChild)
+	join.SetSchema(append(lChild.GetSchema().DeepCopy(), rChild.GetSchema().DeepCopy()...))
+	lChild.SetParents(join)
+	rChild.SetParents(join)
+	return join
+}
+
+func (e *joinReOrderSolver) walkGraph(u int) {
```
Comment (coocood) [inline plan/join_reorder.go]: The function name doesn't reflect what it does.

File: plan/join_reorder.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+			rCol, rok := f.Args[1].(*expression.Column)
+			if lok && rok {
+				lID := findColumnIndexByGroup(groups, lCol)
+				rID := findColumnIndexByGroup(groups, rCol)
+				e.graph[lID] = append(e.graph[lID], rID)
+				e.graph[rID] = append(e.graph[rID], lID)
+			}
+		}
+	}
+	var crossJoinGroup []LogicalPlan
```
Comment (coocood) [inline plan/join_reorder.go]: Do you mean Cartesian product join?

File: plan/logical_plans.go [hunk 4]
```diff
@@ -25,10 +25,8 @@ import (
 type JoinType int
 
 const (
-	// CrossJoin means Cartesian Product, but not used now.
-	CrossJoin JoinType = iota
 	// InnerJoin means inner join.
-	InnerJoin
+	InnerJoin = iota
```
Comment (shenli) [inline plan/logical_plans.go]: InnerJoin JoinType = iota

File: plan/physical_plan_builder.go [hunk 5]
```diff
@@ -206,14 +210,22 @@ func addPlanToResponse(p PhysicalPlan, res *physicalPlanInfo) *physicalPlanInfo
 
 // convert2PhysicalPlan implements LogicalPlan convert2PhysicalPlan interface.
 func (p *Limit) convert2PhysicalPlan(prop requiredProperty) (*physicalPlanInfo, *physicalPlanInfo, uint64, error) {
-	res0, res1, count, err := p.GetChildByIndex(0).(LogicalPlan).convert2PhysicalPlan(prop)
+	res0, res1, count := p.getPlanInfo(prop)
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Why not use sortedRes, unsortedRes?

File: plan/physical_plan_builder.go [hunk 6]
```diff
@@ -206,14 +210,22 @@ func addPlanToResponse(p PhysicalPlan, res *physicalPlanInfo) *physicalPlanInfo
...
+	var err error
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Move this line after the following if statement.

File: plan/physical_plan_builder.go [hunk 7]
```diff
@@ -303,6 +315,10 @@ func (p *Join) handleRightJoin(prop requiredProperty, innerJoin bool) (*physical
 
 // convert2PhysicalPlan implements LogicalPlan convert2PhysicalPlan interface.
 func (p *Join) convert2PhysicalPlan(prop requiredProperty) (*physicalPlanInfo, *physicalPlanInfo, uint64, error) {
+	r0, r1, cnt := p.getPlanInfo(prop)
```
Comment (shenli) [inline plan/physical_plan_builder.go]: It is better to make the var name consistent.

File: plan/plan.go [hunk 8]
```diff
@@ -152,9 +152,33 @@ type PhysicalPlan interface {
 }
 
 type baseLogicalPlan struct {
+	propLen  int
+	propRes0 *physicalPlanInfo
```
Comment (shenli) [inline plan/plan.go]: Not good member variable names.

File: plan/join_reorder.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+			cols, _ := extractColumn(f, nil, nil)
+			for _, col := range cols {
+				idx := findColumnIndexByGroup(group, col)
+				if id == -1 {
+					switch f.FuncName.L {
+					case ast.EQ:
+						rate *= 0.1
+					case ast.LT, ast.LE, ast.GE, ast.GT:
+						rate *= 0.3
+					default:
```
Comment (coocood) [inline plan/join_reorder.go]: For condition like `a = 1 or b = 1`, this algorithm is not good.

File: plan/join_reorder.go [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+					id = -1
+					break
+				}
+			}
+			if id != -1 {
+				e.groupRank[id].rate *= rate
+			}
+		}
+	}
+	for _, edge := range e.graph {
```
Comment (coocood) [inline plan/join_reorder.go]: The element in a graph should be node rather than edge.

File: plan/join_reorder.go [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+	}
+	join.initID()
+	join.SetChildren(lChild, rChild)
+	join.SetSchema(append(lChild.GetSchema().DeepCopy(), rChild.GetSchema().DeepCopy()...))
+	lChild.SetParents(join)
+	rChild.SetParents(join)
+	return join
+}
+
+// walkGraph implement a dfs algorithm. Each time it picks a edge with lowest rate.
```
Comment (coocood) [inline plan/join_reorder.go]: The logic `Each time it picks an edge with lowest rate` is not implemented in this function, this function doesn't care about `rate`.

Comment (hanfei19910905) [inline plan/join_reorder.go]: It really cares, because the edge is sorted.

Comment (coocood) [inline plan/join_reorder.go]: It's not sorted in this function.

File: plan/join_reorder.go [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+import (
+	"sort"
+
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+)
+
+// tryToGetJoinGroup try to fetch a whole join group, which all joins is cartesian join.
```
Comment (zimulala) [inline plan/join_reorder.go]: try -> tries?

File: plan/join_reorder.go [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+func (e *joinReOrderSolver) Swap(i, j int) {
+	e.groupRank[i], e.groupRank[j] = e.groupRank[j], e.groupRank[i]
+}
+
+func (e *joinReOrderSolver) Len() int {
+	return len(e.groupRank)
+}
+
+// reorderJoin implements a simple join reorder algorithm. It will extract all the equal conditions and compose them to a graph.
+// Then walk through the graph and pick the nodes connected by some edge to compose a join tree.
```
Comment (zimulala) [inline plan/join_reorder.go]: edge -> edges

File: plan/join_reorder.go (hunk 11 above)
Comment (zimulala) [inline plan/join_reorder.go]: implement -> implements

File: plan/join_reorder.go (hunk 9 above)
Comment (hanfei19910905) [inline plan/join_reorder.go]: I will add this to todo list

**Comment by coocood** [comment]
LGTM

File: server/conn.go [hunk 14]
```diff
@@ -382,7 +382,11 @@ func (cc *clientConn) handleQuery(sql string) (err error) {
 		err = cc.writeOK()
 	}
 	costTime := time.Now().Sub(startTs)
-	log.Debugf("[TIME_QUERY] %v %s", costTime, sql)
+	if costTime < time.Second {
+		log.Infof("[TIME_QUERY] %v %s", costTime, sql)
```
Comment (shenli) [inline server/conn.go]: Debug is enough.

**Comment by shenli** [comment]
LGTM

**Pull request merged**